### PR TITLE
2229: Fix diagnostic issue when using FakeReads in closures

### PR DESCRIPTION
**Pull request by roxelo**
This PR fixes a diagnostic issue caused by https://github.com/rust-lang/rust/pull/82536. A temporary work around was used in this merged PR which involved feature gating the addition of FakeReads introduced as a result of pattern matching in closures.

The fix involves adding an optional closure DefId to ForLet and ForMatchedPlace FakeReadCauses. This DefId will only be added if a closure pattern matches a Place starting with an Upvar.

r? @nikomatsakis

### CODE REVIEW

**Comment by nikomatsakis** [comment]
@roxelo is there a test case for this diagnostic issue?

**Comment by roxelo** [comment]
Yes, the test case that should pass without feature gate is `src/test/ui/pattern/move-ref-patterns/borrowck-move-ref-pattern.rs`.
It previously used to output this instead: https://github.com/rust-lang/rust/pull/82536#discussion_r591959353.

**Comment by nikomatsakis** [comment]
I see, ok!

File: compiler/rustc_middle/src/mir/mod.rs [hunk 1]
```diff
@@ -1470,7 +1470,7 @@ pub struct Statement<'tcx> {
 
 // `Statement` is used a lot. Make sure it doesn't unintentionally get bigger.
 #[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
-static_assert_size!(Statement<'_>, 32);
+static_assert_size!(Statement<'_>, 40);
```
Comment (nikomatsakis) [inline compiler/rustc_middle/src/mir/mod.rs:1473]: Hmm, this might be a bit of an issue

Comment (nikomatsakis) [reply compiler/rustc_middle/src/mir/mod.rs:1473]: We're making all statements bigger here which seems non-ideal.

Comment (nikomatsakis) [reply compiler/rustc_middle/src/mir/mod.rs:1473]: We could refactor `FakeRead` to have a `Box<FakeReadData>` which contains the `FakeReadCause` and the `Place<'tcx>` (no need to box it separately)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: compiler/rustc_middle/src/mir/mod.rs (hunk 1 above)
Comment (roxelo) [reply compiler/rustc_middle/src/mir/mod.rs]: I added a `box (FakeReadCause, Place<'tcx>)`

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@roxelo looks like this needs another rebase :(

Also, keep in mind that you can instruct bors with r=nikomatsakis afterwards.

@bors delegate+

**Comment by roxelo** [comment]
r=nikomatsakis

**Pull request merged**